In [1]:
import pandas as pd

df = pd.read_csv('reviews.csv')
display(df.head())

id       asins   brand                  categories  \
0  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
1  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
2  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
3  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   
4  AVpe7AsMilAPnD_xQ78G  B00QJDU3KY  Amazon  Amazon Devices,mazon.co.uk   

  colors             dateAdded           dateUpdated  \
0    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
1    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
2    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
3    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   
4    NaN  2016-03-08T20:21:53Z  2017-07-18T23:52:58Z   

                  dimension  ean                         keys  ...  \
0  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
1  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
2  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
3  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   
4  169 mm x 117 mm x 9.1 mm  NaN  kindlepaperwhite/b00qjdu3ky  ...   

  reviews.rating                                 reviews.sourceURLs  \
0            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
1            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
2            4.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
3            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   
4            5.0  https://www.amazon.com/Kindle-Paperwhite-High-...   

                                        reviews.text  \
0  I initially had trouble deciding between the p...   
1  Allow me to preface this with a little history...   
2  I am enjoying it so far. Great for reading. Ha...   
3  I bought one of the first Paperwhites and have...   
4  I have to say upfront - I don't like coroporat...   

                                reviews.title reviews.userCity  \
0              Paperwhite voyage, no regrets!              NaN   
1           One Simply Could Not Ask For More              NaN   
2  Great for those that just want an e-reader              NaN   
3                    Love / Hate relationship              NaN   
4                                   I LOVE IT              NaN   

  reviews.userProvince    reviews.username  sizes upc     weight  
0                  NaN          Cristina M    NaN NaN  205 grams  
1                  NaN               Ricky    NaN NaN  205 grams  
2                  NaN       Tedd Gardiner    NaN NaN  205 grams  
3                  NaN              Dougal    NaN NaN  205 grams  
4                  NaN  Miljan David Tanic    NaN NaN  205 grams  

[5 rows x 27 columns]

In [2]:
print(df.columns)

Index(['id', 'asins', 'brand', 'categories', 'colors', 'dateAdded',
       'dateUpdated', 'dimension', 'ean', 'keys', 'manufacturer',
       'manufacturerNumber', 'name', 'prices', 'reviews.date',
       'reviews.doRecommend', 'reviews.numHelpful', 'reviews.rating',
       'reviews.sourceURLs', 'reviews.text', 'reviews.title',
       'reviews.userCity', 'reviews.userProvince', 'reviews.username', 'sizes',
       'upc', 'weight'],
      dtype='object')


In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1597 entries, 0 to 1596
Data columns (total 27 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    1597 non-null   object 
 1   asins                 1597 non-null   object 
 2   brand                 1597 non-null   object 
 3   categories            1597 non-null   object 
 4   colors                774 non-null    object 
 5   dateAdded             1597 non-null   object 
 6   dateUpdated           1597 non-null   object 
 7   dimension             565 non-null    object 
 8   ean                   898 non-null    float64
 9   keys                  1597 non-null   object 
 10  manufacturer          965 non-null    object 
 11  manufacturerNumber    902 non-null    object 
 12  name                  1597 non-null   object 
 13  prices                1597 non-null   object 
 14  reviews.date          1217 non-null   object 
 15  reviews.doRecommend  

In [9]:
df_subset = df[['name', 'reviews.text', 'reviews.title', 'reviews.userCity', 'manufacturer', 'weight']].copy()
display(df_subset.head())

name                                       reviews.text  \
0  Kindle Paperwhite  I initially had trouble deciding between the p...   
1  Kindle Paperwhite  Allow me to preface this with a little history...   
2  Kindle Paperwhite  I am enjoying it so far. Great for reading. Ha...   
3  Kindle Paperwhite  I bought one of the first Paperwhites and have...   
4  Kindle Paperwhite  I have to say upfront - I don't like coroporat...   

                                reviews.title  reviews.userCity manufacturer  \
0              Paperwhite voyage, no regrets!               NaN       Amazon   
1           One Simply Could Not Ask For More               NaN       Amazon   
2  Great for those that just want an e-reader               NaN       Amazon   
3                    Love / Hate relationship               NaN       Amazon   
4                                   I LOVE IT               NaN       Amazon   

      weight  
0  205 grams  
1  205 grams  
2  205 grams  
3  205 grams  
4  205 grams

# Task
Load "reviews.csv" into a pandas DataFrame, display columns, describe columns, show column types, show an example value for the 'name' column, find distinct values for the 'name' column, keep only the columns 'name', 'reviews.text', 'reviews.title', 'reviews.userCity', 'manufacturer', and 'weight', and fill any null values in the resulting DataFrame.

## Identify null values

### Subtask:
Determine which columns in the `df_subset` DataFrame have null values and the extent of missingness in each of those columns.


**Reasoning**:
Determine the number of null values in each column of the subset DataFrame.



In [10]:
display(df_subset.isnull().sum())

name                   0
reviews.text           0
reviews.title         17
reviews.userCity    1597
manufacturer         632
weight               911
dtype: int64

**Reasoning**:
The previous command failed because `df_subset` was not defined in the current session. Recreate `df_subset` and then check for null values.



In [11]:
df_subset = df[['name', 'reviews.text', 'reviews.title', 'reviews.userCity', 'manufacturer', 'weight']].copy()
display(df_subset.isnull().sum())

name                   0
reviews.text           0
reviews.title         17
reviews.userCity    1597
manufacturer         632
weight               911
dtype: int64

## Determine strategy for filling nulls

### Subtask:
Based on the types of data in the columns with nulls and the amount of missing data, propose different strategies for filling the nulls (e.g., filling with a placeholder, using a statistical measure like mean/median/mode, or using forward/backward fill).


**Reasoning**:
Analyze the data types and null counts for the relevant columns to propose filling strategies.



In [12]:
print(df_subset.dtypes)
print(df_subset.isnull().sum())

name                 object
reviews.text         object
reviews.title        object
reviews.userCity    float64
manufacturer         object
weight               object
dtype: object
name                   0
reviews.text           0
reviews.title         17
reviews.userCity    1597
manufacturer         632
weight               911
dtype: int64


## Implement null filling

### Subtask:
Generate and execute the code to fill the null values in the `df_subset` DataFrame based on the chosen strategy.


**Reasoning**:
Drop the 'reviews.userCity' column and fill the null values in the specified columns with placeholder strings as determined in the previous step.



In [13]:
df_subset = df_subset.drop(columns=['reviews.userCity'])
df_subset['reviews.title'] = df_subset['reviews.title'].fillna('Unknown Title')
df_subset['manufacturer'] = df_subset['manufacturer'].fillna('Unknown Manufacturer')
df_subset['weight'] = df_subset['weight'].fillna('Unknown Weight')
display(df_subset.isnull().sum())

name             0
reviews.text     0
reviews.title    0
manufacturer     0
weight           0
dtype: int64

## Verify null filling

### Subtask:
Check if the null values have been successfully filled.


**Reasoning**:
Check if the null values have been successfully filled by counting the remaining nulls in each column of the df_subset DataFrame.



In [14]:
display(df_subset.isnull().sum())

name             0
reviews.text     0
reviews.title    0
manufacturer     0
weight           0
dtype: int64

## Summary:

### Data Analysis Key Findings

*   The initial `df_subset` DataFrame contained null values in the following columns: `reviews.title` (17 nulls), `reviews.userCity` (1597 nulls), `manufacturer` (632 nulls), and `weight` (911 nulls).
*   The `reviews.userCity` column consisted entirely of null values.
*   The null values in `reviews.title`, `manufacturer`, and `weight` were filled with placeholder strings: 'Unknown Title', 'Unknown Manufacturer', and 'Unknown Weight' respectively.
*   The `reviews.userCity` column was dropped due to having all null values.
*   After the filling process, all selected columns (`name`, `reviews.text`, `reviews.title`, `manufacturer`, `weight`) had zero null values.

### Insights or Next Steps

*   The 'weight' column is of object type, suggesting inconsistent data formats. A next step could involve examining the unique values and patterns in the 'weight' column to determine if it can be cleaned and converted to a numerical type for potential quantitative analysis.
*   Further analysis could explore the impact of filling null values with placeholders on subsequent analyses involving the 'reviews.title', 'manufacturer', and 'weight' columns.


In [15]:
display(df_subset.head())

name                                       reviews.text  \
0  Kindle Paperwhite  I initially had trouble deciding between the p...   
1  Kindle Paperwhite  Allow me to preface this with a little history...   
2  Kindle Paperwhite  I am enjoying it so far. Great for reading. Ha...   
3  Kindle Paperwhite  I bought one of the first Paperwhites and have...   
4  Kindle Paperwhite  I have to say upfront - I don't like coroporat...   

                                reviews.title manufacturer     weight  
0              Paperwhite voyage, no regrets!       Amazon  205 grams  
1           One Simply Could Not Ask For More       Amazon  205 grams  
2  Great for those that just want an e-reader       Amazon  205 grams  
3                    Love / Hate relationship       Amazon  205 grams  
4                                   I LOVE IT       Amazon  205 grams

In [18]:
df_duplicated = pd.concat([df_subset] * 50, ignore_index=True)
display(f"Original shape: {df_subset.shape}")
display(f"Duplicated shape: {df_duplicated.shape}")

df_duplicated.to_csv('amazon_electronic_reviews.csv', index=False)

'Original shape: (1597, 5)'

'Duplicated shape: (79850, 5)'